### PR TITLE
fix: add improved lograge auto patching support and improve docs

### PR DESCRIPTION
**Pull request by ericmustin**
## Summary

This PR addresses https://github.com/DataDog/dd-trace-rb/issues/1205. Since Rails may load lograge's initializer file sequentially _after_ datadog's (Rails loads initializer's alphabetically and datadog's _suggested config_ filename precedes lograge's _suggested config_ filename), we cannot rely on checking `lograge.enabled` to determine whether to patch lograge with our trace correlation code, as it may yield a false negative.

However, since we can _detect_ if the lograge gem has been included in the application (enabled or otherwise), and since setting `lograge.custom_options` (which we need to set as a lambda that attaches our trace correlation context) is a no-op in the case that `lograge.enabled` is, later, explicitly set to `false` after our patching code, we can "naively" add `custom_options` without impacting the log output in an unexpected way.

## Notes

One edge case we _can't_ handle at the moment is when a user has also defined `custom_options` in their lograge specific initializer. Our ideal behavior is that we merge any existing `custom_options` with our lambda, however we can't do that if our lambda is set first (and get's overwritten by a user's lograge initializer settings). Instead, what we can do is document this edge case and give users possible workarounds (use our manual trace correlation snippet instead). 

Unfortunately, this edge case is not exactly easy to explicitly test in our test suite,  and refactoring the test suite to support testing initializer sequence would be a significant undertaking, imo out of scope of this PR (which will improve reliability).

## Next Steps

This is a more brittle than i'd like, generally speaking. Rails+ Lograge present some unique challenges here. First, there's multiple ways you can configure Lograge logging. I believe the case we're solving here only really occurs with a separate initializer file for the gem (instead of having config for logging live in config/environments/production.rb, which is an option suggested by [Lograge docs too](https://github.com/roidrage/lograge/blob/04b30072063ad0463de198b0bd268881212a8529/README.md#installation) ). And Rails has strong opinions about [initializer file ordering](https://guides.rubyonrails.org/configuring.html#using-initializer-files), and does tell users that they should use file naming manage their initializer ordering. But it's probably not reasonable as a 3rd party Lib to tell users they have to change their filenames to support our features.

If we monkey patched the Lograge gem, maybe [set_lograge_log_options](https://github.com/roidrage/lograge/blob/397c830208a7ec1601c0eacc0dddd09274afff5b/lib/lograge.rb#L184), we can probably have a more robust auto trace correlation with logs, but this requires further investigation.

### CODE REVIEW

**Review by brettlangdon — APPROVED** [review]

File: lib/ddtrace/contrib/rails/patcher.rb [hunk 1]
```diff
@@ -54,14 +54,25 @@ def add_middleware(app)
...
           # if so, add proc that injects trace identifiers for tagged logging.
-          elsif (logger = app.config.logger) && logger.is_a?(::ActiveSupport::TaggedLogging)
+          if (logger = app.config.logger) &&
+             defined?(::ActiveSupport::TaggedLogging) &&
+             logger.is_a?(::ActiveSupport::TaggedLogging)
+
             Datadog::Contrib::Rails::LogInjection.add_as_tagged_logging_logger(app)
           else
-            Datadog.logger.warn("Unabe to enable Datadog Trace context, Logger #{logger} is not supported")
+            Datadog.logger.warn("Unable to enable Datadog Trace context, Logger #{logger} is not supported")
```
Comment (Supy) [inline lib/ddtrace/contrib/rails/patcher.rb:75]: Double check me here, but if makes use of Lograge and not the Tagger Logger, this warning will appear in ones logs, despite Datadog actually supporting **at least one** logger. This line should only be logged if neither injection methods could be used.

Comment (ericmustin) [reply lib/ddtrace/contrib/rails/patcher.rb:75]: yea, that's right 🤦 . The release we just cut was mainly to unblock a resque support issue, so we should have another release coming sooner than usual. We'll make sure to cover this
